### PR TITLE
Feature: Presenter mode

### PR DESCRIPTION
**Pull request by jallamsetty1**
- Adds the ability to share video as a Pip when screenshare is in progress.
- Local video is controlled only by the camera button when screenshare is in progress.
- The resolution for the presenter camera is calculated based on the resolution of the desktop share that is in progress.
- A new media type "presenter" has been added and the newly created video track is added to the redux store. No new presenter APIs have been added.
- This PR has changes only for the presenter mode, the crop effect and renaming of the tensor flow related bundle changes will be in separate PR

### CODE REVIEW

File: react/features/base/conference/middleware.js [hunk 1]
```diff
@@ -589,7 +590,7 @@ function _syncReceiveVideoQuality({ getState }, next, action) {
 function _trackAddedOrRemoved(store, next, action) {
     const track = action.track;
 
-    if (track && track.local) {
+    if (track && track.local && track.mediaType !== MEDIA_TYPE.PRESENTER) {
```
Comment (saghul) [inline react/features/base/conference/middleware.js]: Can you add a TODO note here? In the future all track swapping should probably happen here instead of conference.js, but it's now nearly impossible to do, so let's leave a ccomment as a reminder.

File: react/features/base/media/actions.js [hunk 2]
```diff
@@ -97,6 +102,7 @@ export function setVideoAvailable(available: boolean) {
  */
 export function setVideoMuted(
         muted: boolean,
+        mediaType: MEDIA_TYPE = MEDIA_TYPE.VIDEO,
```
Comment (saghul) [inline react/features/base/media/actions.js]: Why does this need to be explicit? I think this action could figure out what to do based on the current track state, can't it?

File: react/features/base/media/middleware.js [hunk 3]
```diff
@@ -45,7 +49,13 @@ MiddlewareRegistry.register(store => next => action => {
         const result = next(action);
         const { track } = action;
 
-        track.local && _syncTrackMutedState(store, track);
+        // Don't sync track mute state with the redux store for screenshare
+        // since video mute state represents local camera mute state only.
+        // Also, do not try to sync mute state for the presenter track since
```
Comment (saghul) [inline react/features/base/media/middleware.js]: How is that possible? Since we run `next(action)` above the redux store should be up to date already. Can you please clarify?

File: react/features/base/tracks/functions.js [hunk 4]
```diff
@@ -53,16 +53,21 @@ export function createLocalTracksF(
...
+        ?? state['features/base/config'].constraints;
+
+    // Do not load blur effect if option for ignoring effects is present.
+    // This is needed when we are creating a video track for presenter mode.
+    const loadEffectsPromise = !options.ignoreEffects
+        && state['features/blur'].blurEnabled ? getBlurEffect()
             .then(blurEffect => [ blurEffect ])
             .catch(error => {
-                logger.error('Failed to obtain the blur effect instance with error: ', error);
+                logger.error('Failed to obtain the blur effect instance with'
```
Comment (saghul) [inline react/features/base/tracks/functions.js]: unnecessary change?

File: react/features/base/tracks/middleware.js [hunk 5]
```diff
@@ -265,8 +266,12 @@ function _setMuted(store, { ensureTrack, muted }, mediaType: MEDIA_TYPE) {
         // `jitsiTrack`, then the `muted` state will be applied once the
         // `jitsiTrack` is created.
         const { jitsiTrack } = localTrack;
+        const isAudioOnly = authority === VIDEO_MUTISM_AUTHORITY.AUDIO_ONLY;
 
-        jitsiTrack && setTrackMuted(jitsiTrack, muted);
+        // screenshare cannot be muted or unmuted using the video mute button
```
Comment (saghul) [inline react/features/base/tracks/middleware.js]: this is sort of what I meant earlier. Since we can make these checks internally, we can probably drop the media type parameter from setVideoMuted altogether

File: react/features/stream-effects/presenter/JitsiStreamPresenterEffect.js [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+    /**
+     * Checks if the local track supports this effect.
+     *
+     * @param {JitsiLocalTrack} jitsiLocalTrack - Track to apply effect.
+     * @returns {boolean} - Returns true if this effect can run on the
+     * specified track, false otherwise.
+     */
+    isEnabled(jitsiLocalTrack) {
+        return jitsiLocalTrack.isVideoTrack()
+            && jitsiLocalTrack.videoType === 'desktop';
```
Comment (saghul) [inline react/features/stream-effects/presenter/JitsiStreamPresenterEffect.js]: doesn't this fit in a single line?

File: react/features/stream-effects/presenter/index.js [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+/**
+ * Creates a new instance of JitsiStreamPresenterEffect.
+ *
+ * @param {MediaStream} stream - The video stream which will be used for
+ * creating the presenter effect.
+ * @returns {Promise<JitsiStreamPresenterEffect>}
+ */
+export function createPresenterEffect(stream: MediaStream) {
+    if (!MediaStreamTrack.prototype.getSettings
+        && !MediaStreamTrack.prototype.getConstraints) {
```
Comment (saghul) [inline react/features/stream-effects/presenter/index.js]: style: please indent this line one more level

File: react/features/stream-effects/presenter/index.js [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+        return Promise.reject(new Error('JitsiStreamPresenterEffect not '
```
Comment (saghul) [inline react/features/stream-effects/presenter/index.js]: doesn't this fit in a single line?

File: react/features/toolbox/components/VideoMuteButton.js [hunk 9]
```diff
@@ -164,10 +164,14 @@ class VideoMuteButton extends AbstractVideoMuteButton<Props, *> {
...
     const tracks = state['features/base/tracks'];
+    const videoTrack = getLocalVideoTrack(tracks);
+    const mediaType = videoTrack && videoTrack.videoType === 'desktop'
+        ? MEDIA_TYPE.PRESENTER
+        : MEDIA_TYPE.VIDEO;
 
     return {
         _audioOnly: Boolean(audioOnly),
-        _videoMuted: isLocalTrackMuted(tracks, MEDIA_TYPE.VIDEO)
+        _videoMuted: isLocalTrackMuted(tracks, mediaType)
```
Comment (saghul) [inline react/features/toolbox/components/VideoMuteButton.js]: WDYT about adding some `isLocalVideoMuted(tracks)` which does this media type dance?

File: react/features/stream-effects/presenter/JitsiStreamPresenterEffect.js (hunk 6 above)
Comment (jallamsetty1) [reply react/features/stream-effects/presenter/JitsiStreamPresenterEffect.js]: Are we not sticking to 80 characters per line limit ?

File: react/features/stream-effects/presenter/index.js (hunk 8 above)
Comment (jallamsetty1) [reply react/features/stream-effects/presenter/index.js]: This line is also longer than 80 characters.

File: react/features/stream-effects/presenter/index.js (hunk 7 above)
Comment (jallamsetty1) [reply react/features/stream-effects/presenter/index.js]: ok

File: react/features/toolbox/components/VideoMuteButton.js (hunk 9 above)
Comment (jallamsetty1) [reply react/features/toolbox/components/VideoMuteButton.js]: Done, I added isLocalVideoTrackMuted(tracks) method to check for mute status on the presenter track if present, otherwise it returns the mute status of the video track.

File: react/features/base/tracks/functions.js (hunk 4 above)
Comment (jallamsetty1) [reply react/features/base/tracks/functions.js]: Yes, it was 95 characters long but I reverted it now.

File: react/features/base/media/actions.js (hunk 2 above)
Comment (jallamsetty1) [reply react/features/base/media/actions.js]: As discussed with you, we would need this to mute either of the presenter or video tracks explicitly.

File: react/features/base/conference/middleware.js (hunk 1 above)
Comment (jallamsetty1) [reply react/features/base/conference/middleware.js]: Sure

File: conference.js [hunk 10]
```diff
@@ -1532,6 +1680,50 @@ export default {
...
+
+        try {
+            const videoTracks = await createLocalTracksF({
+                constraints,
+                devices: [ 'video' ],
+                cameraDeviceId,
+                ignoreEffects: true
+            });
+
+            videoTracks[0].type = MEDIA_TYPE.PRESENTER;
```
[the remark this answers is not in the document]
Comment (jallamsetty1) [reply conference.js]: Sounds like a good idea. Added a new function createLocalPresenterTrack for creating a presenter track.

File: conference.js [hunk 11]
```diff
@@ -1532,6 +1680,50 @@ export default {
...
+     * currently shared.
+     * @param {string} cameraDeviceId - The device id of the camera to be used.
+     * @return {Promise<JitsiStreamPresenterEffect>} - A promise resolved with
+     * {@link JitsiStreamPresenterEffect} if it succeeds.
+     */
+    async _createPresenterStreamEffect(height, cameraDeviceId = null) {
+        const constraints = this._createCameraConstraints(height);
+
+        try {
+            const videoTracks = await createLocalTracksF({
```
[the remark this answers is not in the document]
Comment (jallamsetty1) [reply conference.js]: ok.

File: conference.js [hunk 12]
```diff
@@ -812,6 +867,11 @@ export default {
             return;
         }
 
+        if (this.isSharingScreen) {
+
```
[the remark this answers is not in the document]
Comment (jallamsetty1) [reply conference.js]: removed the newline

File: conference.js [hunk 13]
```diff
@@ -798,6 +809,50 @@ export default {
...
+     */
+    async mutePresenterVideo(mute, showUI = true) {
+        const maybeShowErrorDialog = error => {
+            showUI && APP.store.dispatch(notifyCameraError(error));
+        };
+
+        if (mute) {
+            return this._turnPresenterModeOff();
+        }
+        try {
```
[the remark this answers is not in the document]
Comment (jallamsetty1) [reply conference.js]: ok

File: conference.js [hunk 14]
```diff
@@ -798,6 +809,50 @@ export default {
...
+            const effect = await this._createPresenterStreamEffect(
+                height, defaultCamera);
+
+            if (!effect) {
+                logger.error('Failed to unmute Presenter Video');
+
+                return;
+            }
+            this.localVideo.setEffect(effect)
+                .then(() => {
```
[the remark this answers is not in the document]
Comment (jallamsetty1) [reply conference.js]: ok

File: conference.js [hunk 15]
```diff
@@ -1429,14 +1524,67 @@ export default {
...
+                const effect = await this._createPresenterStreamEffect(
+                    height, defaultCamera);
+
+                if (!effect) {
+                    logger.error('Failed to create the presenter effect');
+
+                    return Promise.resolve();
+                }
+
+                return this.localVideo.setEffect(effect)
```
[the remark this answers is not in the document]
Comment (jallamsetty1) [reply conference.js]: ok

File: conference.js [hunk 16]
```diff
@@ -1429,14 +1524,67 @@ export default {
...
+                const { height } = this.localVideo.track.getSettings();
+                const defaultCamera
+                    = getUserSelectedCameraDeviceId(APP.store.getState());
+                const effect = await this._createPresenterStreamEffect(
+                    height, defaultCamera);
+
+                if (!effect) {
+                    logger.error('Failed to create the presenter effect');
+
+                    return Promise.resolve();
```
[the remark this answers is not in the document]
Comment (jallamsetty1) [reply conference.js]: ok

File: conference.js [hunk 17]
```diff
@@ -1429,14 +1524,67 @@ export default {
...
+                }
+
+                return this.localVideo.setEffect(effect)
+                    .then(() => {
+                        muteLocalVideo(false);
+                        this.setVideoMuteStatus(false);
+                    });
+            } catch (err) {
+                logger.error('Failed to switch to screensharing');
+                Promise.reject(err);
```
[the remark this answers is not in the document]
Comment (jallamsetty1) [reply conference.js]: ok

File: conference.js [hunk 18]
```diff
@@ -1532,6 +1680,50 @@ export default {
...
+                devices: [ 'video' ],
+                cameraDeviceId,
+                ignoreEffects: true
+            });
+
+            videoTracks[0].type = MEDIA_TYPE.PRESENTER;
+            this.localPresenterVideo = videoTracks[0];
+            APP.store.dispatch(trackAdded(this.localPresenterVideo));
+
+            return createPresenterEffect(videoTracks[0].stream)
```
[the remark this answers is not in the document]
Comment (jallamsetty1) [reply conference.js]: done

File: conference.js [hunk 19]
```diff
@@ -1532,6 +1680,50 @@ export default {
...
+                    APP.store.dispatch(
+                        trackRemoved(this.localPresenterVideo));
+                    APP.store.dispatch(notifyCameraError(err));
+
+                    return Promise.resolve();
+                });
+        } catch (err) {
+            logger.error('Failed to create a camera track for presenter');
+
+            return Promise.resolve();
```
[the remark this answers is not in the document]
Comment (jallamsetty1) [reply conference.js]: Yes I agree.

File: react/features/toolbox/components/VideoMuteButton.js [hunk 20]
```diff
@@ -136,10 +135,13 @@ class VideoMuteButton extends AbstractVideoMuteButton<Props, *> {
             this.props.dispatch(
                 setAudioOnly(false, /* ensureTrack */ true));
         }
+        const tracks = APP.store.getState()['features/base/tracks'];
```
Comment (saghul) [inline react/features/toolbox/components/VideoMuteButton.js]: this won't work on mobile (there is no APP). Extract the mediaType in `_mapStateToProps`

File: conference.js [hunk 21]
```diff
@@ -1390,6 +1447,16 @@ export default {
             promise = this.useVideoStream(null);
         }
 
+        // mute the presenter track if it exists.
+        if (this.localPresenterVideo) {
+            APP.store.dispatch(
+                setVideoMuted(true, MEDIA_TYPE.PRESENTER));
```
Comment (saghul) [inline conference.js]: We might need to add a little something here. When muting the presenter, viewers will see the video flickering from screen + video to nothing back to screen. By looking at the thumbnail I noticed that the receivers *think* the video is completely muted (we send it over siognalling), but we shouldn't in this case.

I'm leaving this comment here because I think it's in context, not necessarily because the fix must happen here.

Comment (jallamsetty1) [reply conference.js]: This because of how setEffect is implemented. We remove the stream (w/ efffect) and then add the same stream (w/o effect) back to the conference. The flickering will go away when we switch to using RTCRtpSender.replaceTrack().

File: react/features/base/tracks/functions.js [hunk 22]
```diff
@@ -197,6 +258,19 @@ export function getTracksByMediaType(tracks, mediaType) {
...
+/**
+ * Checks if the local video track in the given set of tracks is muted.
+ *
+ * @param {Track[]} tracks - List of all tracks.
+ * @returns {Track[]}
+ */
+export function isLocalVideoTrackMuted(tracks) {
+    const presenterTrack = getLocalTrack(tracks, MEDIA_TYPE.PRESENTER);
+
+    return presenterTrack ? isLocalTrackMuted(tracks, MEDIA_TYPE.PRESENTER)
```
[the remark this answers is not in the document]
Comment (jallamsetty1) [reply react/features/base/tracks/functions.js]: The suggested change doesn't work since isLocalTrackMuted returns true when the track is not present. So returns true even though video exists when not screensharing.

File: conference.js [hunk 23]
```diff
@@ -1532,6 +1631,46 @@ export default {
...
+            presenterTrack = await createLocalPresenterTrack({
+                cameraDeviceId
+            },
+            height);
+        } catch (err) {
+            logger.error('Failed to create a camera track for presenter', err);
+
+            return;
+        }
+        presenterTrack.type = MEDIA_TYPE.PRESENTER;
```
[the remark this answers is not in the document]
Comment (jallamsetty1) [reply conference.js]: ok

File: react/features/base/tracks/functions.js [hunk 24]
```diff
@@ -10,6 +10,51 @@ import {
...
+            height => (desktopHeight / proportion) < height);
+    const constraints = {
+        video: {
+            aspectRatio: 4 / 3,
+            height: {
+                exact: result
+            }
+        }
+    };
+    let videoTrack;
```
[the remark this answers is not in the document]
Comment (jallamsetty1) [reply react/features/base/tracks/functions.js]: ok

File: react/features/base/tracks/functions.js [hunk 25]
```diff
@@ -10,6 +10,51 @@ import {
...
+        [ videoTrack ] = await JitsiMeetJS.createLocalTracks(
+            {
+                cameraDeviceId,
+                constraints,
+                devices: [ 'video' ]
+            });
+
+        return videoTrack;
+    } catch (err) {
+        logger.error('Failed to create local track for presenter effect', err);
```
[the remark this answers is not in the document]
Comment (jallamsetty1) [reply react/features/base/tracks/functions.js]: ok

File: react/features/base/tracks/middleware.js [hunk 26]
```diff
@@ -131,7 +132,7 @@ MiddlewareRegistry.register(store => next => action => {
             const { jitsiTrack } = action.track;
             const muted = jitsiTrack.isMuted();
             const participantID = jitsiTrack.getParticipantId();
-            const isVideoTrack = jitsiTrack.isVideoTrack();
+            const isVideoTrack = jitsiTrack.type === MEDIA_TYPE.VIDEO || jitsiTrack.type === MEDIA_TYPE.PRESENTER;
```
[the remark this answers is not in the document]
Comment (jallamsetty1) [reply react/features/base/tracks/middleware.js]: ok

File: conference.js (hunk 21 above)
Comment (saghul) [reply conference.js]: Okidoki!

File: react/features/base/tracks/functions.js (hunk 22 above)
[the remark this answers is not in the document]
Comment (saghul) [reply react/features/base/tracks/functions.js]: Makes sense. Maybe add a comment please, since it isn't obvious at first glance.

[the remark this answers is not in the document]
Comment (jallamsetty1) [reply react/features/base/tracks/functions.js]: Sure, I had to add another check as well.

File: react/features/toolbox/components/VideoMuteButton.js (hunk 20 above)
Comment (jallamsetty1) [reply react/features/toolbox/components/VideoMuteButton.js]: ok, made the change.

File: react/features/base/tracks/functions.js [hunk 27]
```diff
@@ -10,6 +10,52 @@ import {
...
+                cameraDeviceId,
+                constraints,
+                devices: [ 'video' ]
+            });
+
+        videoTrack.type = MEDIA_TYPE.PRESENTER;
+
+        return videoTrack;
+    } catch (err) {
+        return err;
```
Comment (saghul) [inline react/features/base/tracks/functions.js]: remove the try catch instead of returning the error. As it stands, it would resolve the promise succcesfully, but with an error object. We want it to be rejected instead. Not having the try will make an eventual exception reject the promise.

File: react/features/stream-effects/presenter/JitsiStreamPresenterEffect.js [hunk 28]
```diff
@@ -0,0 +1,166 @@
...
+            ?? firstVideoTrack.getConstraints();
+
+        this._canvas = document.createElement('canvas');
+        this._ctx = this._canvas.getContext('2d');
+
+        if (document.body !== null) {
+            document.body.appendChild(this._canvas);
+        }
+        this._desktopElement = document.createElement('video');
+        this._videoElement = document.createElement('video');
```
Comment (saghul) [inline react/features/stream-effects/presenter/JitsiStreamPresenterEffect.js]: set the class to flip it?

**Comment by paweldomas** [comment]
jenkins test this please
